File: app/src/main/java/com/recursia/popularmovies/presentation/views/fragments/DetailScreenFragment.java
package com.recursia.popularmovies.presentation.views.fragments;

import android.content.Intent;
import android.graphics.Matrix;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.chip.Chip;
import android.support.design.chip.ChipGroup;
import android.support.design.widget.CollapsingToolbarLayout;
import android.support.design.widget.FloatingActionButton;
import android.support.v7.widget.CardView;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.arellomobile.mvp.MvpAppCompatFragment;
import com.arellomobile.mvp.presenter.InjectPresenter;
import com.arellomobile.mvp.presenter.ProvidePresenter;
import com.bumptech.glide.Glide;
import com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions;
import com.recursia.popularmovies.R;
import com.recursia.popularmovies.TheApplication;
import com.recursia.popularmovies.di.AppComponent;
import com.recursia.popularmovies.domain.models.Cast;
import com.recursia.popularmovies.domain.models.Genre;
import com.recursia.popularmovies.domain.models.Movie;
import com.recursia.popularmovies.domain.models.Trailer;
import com.recursia.popularmovies.presentation.presenters.DetailScreenPresenter;
import com.recursia.popularmovies.presentation.views.adapters.CastAdapter;
import com.recursia.popularmovies.presentation.views.adapters.MoviesAdapter;
import com.recursia.popularmovies.presentation.views.adapters.TrailersAdapter;
import com.recursia.popularmovies.presentation.views.decorations.MarginItemDecoration;
import com.recursia.popularmovies.presentation.views.contracts.DetailScreenContract;
import com.recursia.popularmovies.utils.DateUtils;
import com.recursia.popularmovies.utils.NetworkUtils;
import com.recursia.popularmovies.utils.TagUtils;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

//TODO return is empty checks back to presenter
public class DetailScreenFragment extends MvpAppCompatFragment implements DetailScreenContract {
    private final static int FADE_OUT_DURATION = 100; //ms
    private final static boolean IS_RECOMMENDATION_MOVIES = true;
    @BindView(R.id.descriptionTextView)
    TextView descriptionTextView;
    @BindView(R.id.ratingTextView)
    TextView ratingTextView;
    @BindView(R.id.originalTitleTextView)
    TextView originalTitleTextView;
    @BindView(R.id.favoriteIcon)
    FloatingActionButton favoriteIcon;
    @BindView(R.id.releaseDateTextView)
    TextView releaseDateTextView;
    @BindView(R.id.recycleViewTrailers)
    RecyclerView recyclerViewTrailers;
    @BindView(R.id.backdropImage)
    ImageView backdropImage;
    @BindView(R.id.toolbar)
    Toolbar toolbar;
    @BindView(R.id.collapsingToolbar)
    CollapsingToolbarLayout collapsingToolbarLayout;
    @BindView(R.id.genresGroup)
    ChipGroup genresGroup;
    @BindView(R.id.recyclerViewCast)
    RecyclerView recyclerViewCast;
    @BindView(R.id.castCardView)
    CardView castCardView;
    @BindView(R.id.descriptionCardView)
    CardView descriptionCardView;
    @BindView(R.id.detailCardView)
    CardView detailCardView;
    @BindView(R.id.recyclerViewMovieRecommendations)
    RecyclerView recyclerViewMovieRecommendations;
    @BindView(R.id.movieRecommendationCardView)
    CardView movieRecommendationCardView;
    @InjectPresenter
    DetailScreenPresenter presenter;
    private TrailersAdapter trailersAdapter;
    private CastAdapter castAdapter;
    private MoviesAdapter moviesAdapter;
    private Movie movie;

    public static DetailScreenFragment getInstance(int movieId) {
        DetailScreenFragment fragment = new DetailScreenFragment();
        Bundle arguments = new Bundle();
        arguments.putInt(TagUtils.MOVIE_ID, movieId);
        fragment.setArguments(arguments);
        return fragment;
    }

    @Override
    public void showFavoriteIcon() {
        favoriteIcon.show();
    }

    @Override
    public void hideFavoriteIcon() {
        favoriteIcon.hide();
    }

    @OnClick(R.id.favoriteIcon)
    public void onFavoriteIconClicked() {
        presenter.onFavoriteIconClicked(movie);
    }

    @OnClick(R.id.backdropImage)
    public void onBackdropImageClicked() {
        presenter.onBackdropImageClicked(movie);
    }

    @Override
    public void setRecommendationMovies(List<Movie> movies) {
        if (!movies.isEmpty()) {
            moviesAdapter.setMovies(movies);
        }
    }

    @Override
    public void hideRecommendationMovies() {
        movieRecommendationCardView.setVisibility(View.GONE);
    }

    @Override
    public void showRecommendationMovies() {
        movieRecommendationCardView.setVisibility(View.VISIBLE);
    }

    @Override
    public void showErrorMessage(String message) {
        Toast.makeText(getContext(), message, Toast.LENGTH_LONG).show();
    }

    @ProvidePresenter
    DetailScreenPresenter providePresenter() {
        int movieId = getArguments().getInt(TagUtils.MOVIE_ID);
        AppComponent app = TheApplication.getInstance().getAppComponent();
        return new DetailScreenPresenter(app.getDetailScreenInteractor(), app.getRouter(), movieId);
    }

    @Override
    public void setFavoriteIconOn() {
        favoriteIcon.setImageDrawable(getContext().getDrawable(R.drawable.ic_favorite_on));
        favoriteIcon.setImageMatrix(new Matrix()); //trick
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == R.id.shareItem) {
            presenter.onShareIconClicked(movie);
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void showMovieAddedMessage() {
        Toast.makeText(getContext(), getString(R.string.added_favorite), Toast.LENGTH_LONG).show();
    }

    @Override
    public void showMovieRemovedMessage() {
        Toast.makeText(getContext(), getString(R.string.removed_favorite), Toast.LENGTH_LONG).show();
    }

    @Override
    public void shareMovie(Movie movie) {
        String shareMessage = getShareMessage(movie);
        Intent intent = new Intent(Intent.ACTION_SEND);
        intent.putExtra(Intent.EXTRA_TEXT, shareMessage);
        intent.setType("text/plain");
        Intent chosenIntent = Intent.createChooser(intent, getString(R.string.share_to));
        startActivity(chosenIntent);
    }

    private String getShareMessage(Movie movie) {
        return String.format(getString(R.string.share_message), movie.getTitle(), NetworkUtils.getBigPosterUrl(movie.getPosterPath()));
    }

    @Override
    public void hideMovieDetail() {
        detailCardView.setVisibility(View.GONE);
        descriptionCardView.setVisibility(View.GONE);
        recyclerViewTrailers.setVisibility(View.GONE);
        castCardView.setVisibility(View.GONE);
    }

    @Override
    public void showMovieDetail() {
        detailCardView.setVisibility(View.VISIBLE);
        descriptionCardView.setVisibility(View.VISIBLE);
        recyclerViewTrailers.setVisibility(View.VISIBLE);
        castCardView.setVisibility(View.VISIBLE);
    }

    @Override
    public void setFavoriteIconOff() {
        favoriteIcon.setImageDrawable(getContext().getDrawable(R.drawable.ic_favorite_off));
        favoriteIcon.setImageMatrix(new Matrix()); //trick
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_detail, container, false);
        ButterKnife.bind(this, view);
        return view;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        //Trailers
        initTrailerRecyclerView();
        initTrailersAdapter();
        //Cast
        initCastRecyclerView();
        initCreditsAdapter();
        //Movie recommendation
        initMovieRecommendationRecyclerView();
        initMovieRecommendationAdapter();

        initToolbar();
        initCollapsingToolbarLayout();
    }

    private void initTrailerRecyclerView() {
        recyclerViewTrailers.setLayoutManager(new LinearLayoutManager(getContext(), LinearLayoutManager.HORIZONTAL, false));
        recyclerViewTrailers.setHasFixedSize(true);
        recyclerViewTrailers.addItemDecoration(
                new MarginItemDecoration(getContext(), 7, 7, 0, 0));
    }

    private void initTrailersAdapter() {
        trailersAdapter = new TrailersAdapter(getContext());
        trailersAdapter.setClickListener(item -> presenter.onTrailerPlayButtonClicked(item));
        recyclerViewTrailers.setAdapter(trailersAdapter);
    }

    private void initCastRecyclerView() {
        recyclerViewCast.setLayoutManager(new LinearLayoutManager(getContext(), LinearLayoutManager.HORIZONTAL, false));
        recyclerViewCast.setHasFixedSize(true);
        recyclerViewCast.addItemDecoration(
                new MarginItemDecoration(getContext(), 20, 0, 0, 0));
    }

    private void initCreditsAdapter() {
        castAdapter = new CastAdapter(getContext());
        recyclerViewCast.setAdapter(castAdapter);
    }

    private void initMovieRecommendationRecyclerView() {
        recyclerViewMovieRecommendations.setLayoutManager(new LinearLayoutManager(getContext(), LinearLayoutManager.HORIZONTAL, false));
        recyclerViewMovieRecommendations.setHasFixedSize(true);
        recyclerViewMovieRecommendations.addItemDecoration(
                new MarginItemDecoration(getContext(), 7, 7, 0, 0));
    }

    private void initMovieRecommendationAdapter() {
        moviesAdapter = new MoviesAdapter(getContext(), IS_RECOMMENDATION_MOVIES);
        moviesAdapter.setClickListener(item -> presenter.onMovieClicked(item));
        recyclerViewMovieRecommendations.setAdapter(moviesAdapter);
    }

    private void initToolbar() {
        toolbar.setNavigationIcon(R.drawable.ic_arrow_back);
        toolbar.setNavigationOnClickListener(v -> presenter.onBackPressed());
        toolbar.inflateMenu(R.menu.detail_menu);
        toolbar.setOnMenuItemClickListener(menuItem -> {
            if (menuItem.getItemId() == R.id.shareItem) {
                presenter.onShareIconClicked(movie);
                return true;
            }
            return false;
        });
    }

    private void initCollapsingToolbarLayout() {
        collapsingToolbarLayout.setExpandedTitleColor(getResources().getColor(R.color.white));
        collapsingToolbarLayout.setCollapsedTitleTextColor(getResources().getColor(R.color.white));
    }

    @Override
    public void openTrailerUrl(Trailer trailer) {
        Intent browserIntent = new Intent(Intent.ACTION_VIEW, Uri.parse(NetworkUtils.TRAILER_BASE_URL + trailer.getKey()));
        startActivity(browserIntent);
    }

    @Override
    public void setMovieDetail(Movie movie) {
        this.movie = movie;
        //Trailers
        List<Trailer> trailers = movie.getTrailers();
        if (!trailers.isEmpty()) {
            trailersAdapter.setTrailers(trailers);
        }
        //Genres
        List<Genre> genres = movie.getGenres();
        if (!genres.isEmpty()) {
            setGenres(genres);
        }
        //Cast
        List<Cast> casts = movie.getCasts();
        if (!casts.isEmpty()) {
            castAdapter.setCast(casts);
        }
        //Trailers
        collapsingToolbarLayout.setTitle(movie.getTitle());
        originalTitleTextView.setText(movie.getOriginalTitle());
        releaseDateTextView.setText(DateUtils.formatDate(movie.getReleaseDate()));
        descriptionTextView.setText(movie.getOverview());
        ratingTextView.setText(Double.toString(movie.getVoteAverage()));
        //Image
        Glide.with(this)
                .load(NetworkUtils.getBigPosterUrl(movie.getBackdropPath()))
                .transition(DrawableTransitionOptions.withCrossFade(FADE_OUT_DURATION))
                .into(backdropImage);
    }

    private void setGenres(List<Genre> genres) {
        //Before set genres remove previous one
        genresGroup.removeAllViews();
        LayoutInflater layoutInflater = LayoutInflater.from(getContext());
        for (Genre genre : genres) {
            Chip chip = (Chip) layoutInflater.inflate(R.layout.genre_chip, genresGroup, false);
            chip.setText(genre.getName());
            genresGroup.addView(chip);
        }
    }

}

File: app/src/main/java/com/recursia/popularmovies/presentation/views/fragments/DetailScreenFragment.kt
package com.recursia.popularmovies.presentation.views.fragments

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.widget.Toolbar
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.viewpager.widget.ViewPager
import butterknife.BindView
import butterknife.ButterKnife
import com.bumptech.glide.Glide
import com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions
import com.google.android.material.tabs.TabLayout
import com.recursia.popularmovies.R
import com.recursia.popularmovies.TheApplication
import com.recursia.popularmovies.domain.models.Movie
import com.recursia.popularmovies.domain.models.enums.MovieStatus
import com.recursia.popularmovies.presentation.presenters.DetailScreenPresenter
import com.recursia.popularmovies.presentation.views.adapters.MoviePagerAdapter
import com.recursia.popularmovies.presentation.views.adapters.MovieStatusesAdapter
import com.recursia.popularmovies.presentation.views.contracts.DetailScreenContract
import com.recursia.popularmovies.presentation.views.decorations.MarginItemDecoration
import com.recursia.popularmovies.utils.DimensionsUtils
import com.recursia.popularmovies.utils.NetworkUtils
import com.recursia.popularmovies.utils.TagUtils
import moxy.MvpAppCompatFragment
import moxy.presenter.InjectPresenter
import moxy.presenter.ProvidePresenter
import net.opacapp.multilinecollapsingtoolbar.CollapsingToolbarLayout

class DetailScreenFragment : MvpAppCompatFragment(), DetailScreenContract {

    @BindView(R.id.backdrop_image)
    lateinit var backdropImage: ImageView

    @BindView(R.id.toolbar)
    lateinit var toolbar: Toolbar

    @BindView(R.id.collapsing_toolbar)
    lateinit var collapsingToolbarLayout: CollapsingToolbarLayout

    @BindView(R.id.text_view_rating)
    lateinit var textViewRating: TextView

    @BindView(R.id.movie_status_recycler_view)
    lateinit var movieStatusRecyclerView: RecyclerView

    @BindView(R.id.tabs)
    lateinit var tabLayout: TabLayout

    @BindView(R.id.view_pager)
    lateinit var viewPager: ViewPager

    @InjectPresenter
    lateinit var presenter: DetailScreenPresenter

    private lateinit var movieStatusesAdapter: MovieStatusesAdapter
    private var movie: Movie? = null
    private var movieId: Int = 0

    override fun showErrorMessage(message: String) {
        Toast.makeText(context, message, Toast.LENGTH_LONG).show()
    }

    @ProvidePresenter
    internal fun providePresenter(): DetailScreenPresenter {
        movieId = arguments!!.getInt(TagUtils.MOVIE_ID)
        val app = TheApplication.getInstance().appComponent
        return DetailScreenPresenter(app!!.detailScreenInteractor, app.router, movieId)
    }

    override fun shareMovie(movie: Movie) {
        val shareMessage = getShareMessage(movie)
        val intent = Intent(Intent.ACTION_SEND)
        intent.putExtra(Intent.EXTRA_TEXT, shareMessage)
        intent.type = "text/plain"
        val chosenIntent = Intent.createChooser(intent, getString(R.string.share_to))
        startActivity(chosenIntent)
    }

    private fun getShareMessage(movie: Movie): String {
        return String.format(getString(R.string.share_message), movie.title, NetworkUtils.getBigPosterUrl(movie.posterPath!!))
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_detail, container, false)
        ButterKnife.bind(this, view)
        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        // Movie status
        initMovieStatusRecyclerView()
        initMovieStatusAdapter()

        initToolbar()
        initCollapsingToolbarLayout()

        initViewPager()
        initTabLayout()

        // Init listeners
        initOnBackDropClickListener()
    }

    private fun initOnBackDropClickListener() {
        backdropImage.setOnClickListener {
            presenter.onBackdropClicked(movie)
        }
    }

    private fun initTabLayout() {
        tabLayout.setupWithViewPager(viewPager)
    }

    private fun initViewPager() {
        val pagerAdapter = MoviePagerAdapter(childFragmentManager, context!!, movieId)
        viewPager.adapter = pagerAdapter
        viewPager.pageMargin = DimensionsUtils.convertDpToPixel(resources.getDimension(R.dimen.view_pager_margin), context!!).toInt()
    }

    private fun initMovieStatusAdapter() {
        movieStatusesAdapter = MovieStatusesAdapter(context!!)
        movieStatusesAdapter.setOnClickListener {
            presenter.onMovieStatusClicked(movie, it)
        }
        movieStatusRecyclerView.adapter = movieStatusesAdapter
    }

    private fun initMovieStatusRecyclerView() {
        movieStatusRecyclerView.layoutManager = LinearLayoutManager(context, LinearLayoutManager.HORIZONTAL, false)
        movieStatusRecyclerView.addItemDecoration(
                MarginItemDecoration(context!!, 6, 6, 0, 0)
        )
    }

    override fun setMovieStatus(status: MovieStatus) {
        movieStatusesAdapter.setStatusHighlighted(status)
    }

    private fun initToolbar() {
        toolbar.setNavigationIcon(R.drawable.ic_arrow_back)
        toolbar.setNavigationOnClickListener { presenter.onBackPressed() }
        toolbar.inflateMenu(R.menu.detail_menu)
        toolbar.setOnMenuItemClickListener { menuItem ->
            if (menuItem.itemId == R.id.share_item) {
                presenter.onShareIconClicked(movie)
                return@setOnMenuItemClickListener true
            }
            false
        }
    }

    private fun initCollapsingToolbarLayout() {
        collapsingToolbarLayout.setExpandedTitleColor(resources.getColor(R.color.white))
        collapsingToolbarLayout.setCollapsedTitleTextColor(resources.getColor(R.color.white))
        // TODO make it all dimens
        collapsingToolbarLayout.setExpandedTitleMargin(
                DimensionsUtils.convertDpToPixel(22.0f, context!!).toInt(),
                0,
                DimensionsUtils.convertDpToPixel(44.0f, context!!).toInt(),
                DimensionsUtils.convertDpToPixel(75.0f, context!!).toInt()
        )
    }

    override fun setMovieDetail(movie: Movie) {
        this.movie = movie
        collapsingToolbarLayout.title = movie.title
        textViewRating.text = movie.voteAverage.toString()
        movieStatusesAdapter.setStatusHighlighted(movie.status)

        movie.posterPath?.let {
            Glide.with(this)
                    .load(NetworkUtils.getBigPosterUrl(it))
                    .transition(DrawableTransitionOptions.withCrossFade(FADE_OUT_DURATION))
                    .into(backdropImage)
        }
    }

    companion object {
        private const val FADE_OUT_DURATION = 100 // ms

        fun getInstance(movieId: Int): DetailScreenFragment {
            val fragment = DetailScreenFragment()
            val arguments = Bundle()
            arguments.putInt(TagUtils.MOVIE_ID, movieId)
            fragment.arguments = arguments
            return fragment
        }
    }
}
